compiler/src/iree/compiler/Codegen/Common/TransformDialectJitterPass.cpp

```diff
@@ -391,31 +448,114 @@ class TransformDialectJitterPass
     linalg::registerTransformDialectExtension(registry);
   }
 
-  TransformDialectJitterPass() = default;
+  LogicalResult buildStrategy(ImplicitLocOpBuilder ib, Value variantH,
+                              StringRef target, StringRef strategyName,
+                              ArrayRef<int64_t> workgroupSize) {
+    // TODO: this is very brittle.
+    if (target == "cuda") {
+      if (strategyName == "reduction_strategy")
+        return buildReductionCudaStrategy(ib, variantH, workgroupSize);
+    } else if (target == "llvm-cpu") {
+      if (strategyName == "reduction_strategy")
+        return buildReductionCpuStrategy(ib, variantH, workgroupSize);
+    }
+    return getOperation()->emitError("Unknown strategy: ") << strategyName;
+  }
+
+  // TODO: this is very brittle.
+  FailureOr<StringRef> getTarget(HAL::ExecutableVariantOp variantOp) {
+    HAL::ExecutableTargetAttr targetAttr = variantOp.getTarget();
+    if (!targetAttr || !targetAttr.getBackend()) return failure();
+    return targetAttr.getBackend().getValue();
+  }
+
+  // TODO: this is very brittle.
+  FailureOr<HAL::ExecutableExportOp> getAssumedSingleExport(
+      HAL::ExecutableVariantOp variantOp) {
+    HAL::ExecutableExportOp res;
+    auto lambda = [&](HAL::ExecutableExportOp op) {
+      if (res) return WalkResult::interrupt();
+      res = op;
+      return WalkResult::advance();
+    };
+    if (variantOp->walk(lambda).wasInterrupted()) return failure();
+    return res;
+  }
+
+  // TODO: this is very brittle.
+  FailureOr<StringRef> getStrategyName(HAL::ExecutableExportOp exportOp) {
```

Comment:
THese too you can reuse methods from here https://github.com/iree-org/iree/blob/7b2e4ef995305b56776cfc2b8c7a1dec7f2ff176/compiler/src/iree/compiler/Codegen/Dialect/LoweringConfig.h#L124